sympy/series/limits.py

```diff
@@ -68,18 +68,7 @@ def limit(e, z, z0, dir="+"):
      limit_seq : returns the limit of a sequence.
     """
 
-    if dir == "+-":
```

Comment:
Could these lines be on top of `.doit`? That might give a result even when `gruntz` fails.

Reply:
I don't quite understand, if `gruntz` fails it falls through to `heuristics` which just winds up calling `limit` which winds up calling `Limit.doit` which winds up calling `gruntz` term-wise anyway so ... ?

Reply:
Yes, but it would call `Limit.doit` twice, once for `+` and once for `-`, hopefully without calling `gruntz`. It is an expensive method that is intended for those cases where two-sided limits do not exist. In fact, most two-sided limits that exist should be computable with core methods without `gruntz`.

Reply:
From what I see those cases where the limit is computed without using `gruntz` are something like polynomials where the variable goes to infinity, the direction is not taken into consideration in those cases so there is nothing to test for. The only place where the direction actually seems to matter is for calling `gruntz`. What I am saying is that a direction test before this point would be ... well ... pointless.

Reply:
> where two-sided limits exist and are computed using core methods that is already being done without taking the direction into consideration.

Ideally, `doit` should check this before passing over to more sophisticated methods. There should be a core method for something like `is_meromorphic_at`. (At a pole, the two-sided limit would be `zoo`.)

Reply:
> Ideally, `doit` should check this before passing over to more sophisticated methods. There should be a core method for something like `is_meromorphic_at`. (At a pole, the two-sided limit would be `zoo`.)

So, basically, for example: return `zoo` for `limit (1/x, x, 0)` but `oo` for `limit (1/x**2, x, 0)`?

Going by the fact that the direction parameter is only specified as + or - makes me think the current limit implementation was only meant for the domain of the reals, even though it seems to calculate simple complex limits fine.

Is it always the case that if the limit from +/- is `+/-oo` then the complex limit is `zoo`? Or what about the case where the limit is finite but different from the two directions?

Reply:
> the current limit implementation was only meant for the domain of the reals

It looks like this is how the implementation was originally created but, as you point out, it would often be feasible to consider complex limits. In fact, the complex limit usually exists it the two-sided limit exists. It could be possible to extend the meaning of `dir='+-'` to include the complex case.

> Is it always the case that if the limit from +/- is +/-oo then the complex limit is zoo?

I'm not sure about this. It might be possible to have `limit(c/x**2, x, 0, dir='+-') == -oo` if `c < 0`. But it would be simpler to make the limit `zoo`, in particular if we want `dir='+-'` to include complex limit.

> what about the case where the limit is finite but different from the two directions?

In this case the function probably has an essential singularity so there is no limit, neither '+-' nor complex. (In addition, the function behaves very badly: it comes arbitrarily close to any complex value in any neighbourhood by the [Casorati-Weierstrass theorem](https://en.wikipedia.org/wiki/Casorati%E2%80%93Weierstrass_theorem).)

Reply:
So you've identified something which needs to be implemented - complex limits. I remember when I learned about those that the process to evaluate one is closer to a mathematical proof than a computation. I might look into this if I have time, in the meantime this small fix is something which makes `limit` and `Limit` behave consistently so I would say incorporate it and leave the complex limit for another PR.

> I'm not sure about this. It might be possible to have limit(c/x**2, x, 0, dir='+-') == -oo if c < 0. But it would be simpler to make the limit zoo, in particular if we want dir='+-' to include complex limit.

I would say leave the '+-' for limits in the realm of the reals as a bi-directional limit because if you are just doing real math you want `limit (1/x**2, x, 0, '+-')` to exist and be a real infinity and not a complex one. For specifying complex limits I would say a different token altogether, maybe 'c' or 'z'. 

I don't know that much about the topic but it would seem to me that any pole of a complex function in the complex plane would be a complex infinity and not just the essential singularities. Since a pole of a certain order is gotten by dividing by a function of that order then different directions of approach raised to that order would give different complex arguments of the result as you approach z0 and so a complex infinity, no?